### PR TITLE
[MCC-202655] Local Tracing for Arbitrary Events

### PR DESCRIPTION
**Pull request by lschreck-mdsol**
@bvillanueva-mdsol @kenyamat @jcarres-mdsol This PR implements a tracing mechanism for being able to trace local events related to a request.

The concept is the following:
- The client creates a trace which will result an initialized Span
- Calling the newly added ZipkinClient.Record method With the previously initialized span the client is able to save a custom annotation in the current span with the timestamp of the calling
  - For a bonus if there is no value provided for the annotation, the tracer will record the caller method's name as the annotation value
- Another option is to call the ZipkinClient.RecordBinary method which will add a binary annotation to the current span (this can be used to record some general information regarding the context and environment)

Please note that I used some new features of the C# 6.0 (e.g. null-conditional operator, interpolated strings etc.) therefore it is highly recommended to use Visual Studio 2015 for development and build purposes from now on.

Please review and merge it, then I will create and publish the NuGet package.

fyi @cabbott @BPONTES @jfeltesse-mdsol

### CODE REVIEW

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs [hunk 1]
```diff
@@ -69,15 +67,15 @@ public Span StartClientTrace(Uri remoteUri, string methodName)
...
-        }
 
         public void EndClientTrace(Span clientSpan, int statusCode)
         {
-            if (isTraceOn)
-            {
+            if (!isTraceOn)
+                return;
+
                 try
```
Comment (kenyamat) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: Indents are broken.

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs [hunk 2]
```diff
@@ -122,6 +120,55 @@ public void EndServerTrace(Span serverSpan)
...
+            if (!isTraceOn)
+                return;
+
+            try
+            {
+                spanTracer.RecordBinary<T>(span, key, value);
+            }
+            catch (Exception ex)
+            {
+                logger.Error($"Error recording a binary annotation (key: {key})", ex);
```
Comment (kenyamat) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: I got a build error

Comment (kenyamat) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: I use VS2013 maybe it's not problem on VS2015

**Comment by kenyamat** [comment]
LGTM

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs (hunk 2 above)
Comment (lschreck-mdsol) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: I built and ran Ampridatvir with VS2013 and the ZipkinTracerModule built in VS2015 and there were no errors, so this change should impact the development of the ZipkinTracerModule only. At runtime or the development of other applications which referencing this should not have any problems even if VS2013 is used for them.

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs (hunk 1 above)
Comment (lschreck-mdsol) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: Thanks, I have to practice my conflict resolution skills :sweat_smile:. Fixed now.

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs [hunk 3]
```diff
@@ -19,109 +20,155 @@ public class ZipkinClient : ITracerClient
...
-
-                    this.traceProvider = traceProvider;
-                }
-                catch (Exception ex)
-                {
-                    logger.Error("Error Building Zipkin Client Provider", ex);
-                    isTraceOn = false;
-                }
+                logger.Error("Error Building Zipkin Client Provider", ex);
+                isTraceOn = false;
```
Comment (jfeltesse-mdsol) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: why is this needed?

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs (hunk 2 above)
Comment (kenyamat) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: Got it, thanks.

File: src/Medidata.ZipkinTracer.Core/ZipkinClient.cs (hunk 3 above)
Comment (lschreck-mdsol) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: As far as I understand the SpanTracer does the actual trace management (starting, completing, adding annotations etc.) and span composition and the SpanCollector is responsible for the collection and delivery management of the traces.

Maybe @bvillanueva-mdsol can tell you in more detail about this part.

Comment (bvillanueva-mdsol) [inline src/Medidata.ZipkinTracer.Core/ZipkinClient.cs]: @jfeltesse-mdsol @lschreck-mdsol Yes, it will control subsequent calls on different methods. If the the spantracer and spancollector failed to initialized correctly then this flag will prevent any zipkin methods to execute.
